src/manifest/update_periods.ts

```diff
@@ -104,27 +104,38 @@ export function updatePeriods(
                       newPeriods[0],
                       MANIFEST_UPDATE_TYPE.Partial);
 
-  let prevIndexOfNewPeriod = indexOfNewFirstPeriod;
+  let prevIndexOfNewPeriod = indexOfNewFirstPeriod + 1;
   for (let i = 1; i < newPeriods.length; i++) {
     const newPeriod = newPeriods[i];
-    const indexOfNewPeriod = arrayFindIndex(oldPeriods,
-                                            ({ id }) => id === newPeriod.id);
+    let indexOfNewPeriod = -1;
+    for (let j = prevIndexOfNewPeriod; j < oldPeriods.length; j++) {
+      if (newPeriod.id === oldPeriods[j].id) {
+        indexOfNewPeriod = j;
+        break; // end the loop
+      }
+    }
     if (indexOfNewPeriod < 0) {
       oldPeriods.splice(prevIndexOfNewPeriod,
                         oldPeriods.length - prevIndexOfNewPeriod,
                         ...newPeriods.slice(i, newPeriods.length));
       return;
     }
 
+    if (indexOfNewPeriod > prevIndexOfNewPeriod) {
+      oldPeriods.splice(prevIndexOfNewPeriod,
+                        indexOfNewPeriod - prevIndexOfNewPeriod);
+      indexOfNewPeriod = prevIndexOfNewPeriod;
+    }
+
     // Later Periods can be fully replaced
     updatePeriodInPlace(oldPeriods[indexOfNewPeriod],
                         newPeriod,
                         MANIFEST_UPDATE_TYPE.Full);
-    prevIndexOfNewPeriod = indexOfNewPeriod;
+    prevIndexOfNewPeriod++;
   }
 
-  if (prevIndexOfNewPeriod < oldPeriods.length - 1)  {
-    oldPeriods.splice(prevIndexOfNewPeriod + 1,
-                      oldPeriods.length - (prevIndexOfNewPeriod + 1));
+  if (prevIndexOfNewPeriod < oldPeriods.length)  {
+    oldPeriods.splice(prevIndexOfNewPeriod,
+                      oldPeriods.length - prevIndexOfNewPeriod);
   }
 }
```

Comment:
More and more similarities between the replacePeriods and updatePeriods. We may try to merge them into a unique function when we will have a clearer mind.

Reply:
... We already talked about this too much (like 3 tries) without succeeding once.